src/core/qgsmatrix4x4.cpp

```diff
@@ -0,0 +1,166 @@
+/***************************************************************************
+  qgsmatrix4x4.cpp
+  --------------------------------------
+  Date                 : July 2023
+  Copyright            : (C) 2023 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmatrix4x4.h"
+
+// the implementation is partially based on Qt's QMatrix4x4 (simplified)
+
+
+QgsMatrix4x4::QgsMatrix4x4( double m11, double m12, double m13, double m14,
+                            double m21, double m22, double m23, double m24,
+                            double m31, double m32, double m33, double m34,
+                            double m41, double m42, double m43, double m44 )
+{
+  m[0][0] = m11; m[0][1] = m21; m[0][2] = m31; m[0][3] = m41;
+  m[1][0] = m12; m[1][1] = m22; m[1][2] = m32; m[1][3] = m42;
+  m[2][0] = m13; m[2][1] = m23; m[2][2] = m33; m[2][3] = m43;
+  m[3][0] = m14; m[3][1] = m24; m[3][2] = m34; m[3][3] = m44;
+}
+
+
+QgsVector3D operator*( const QgsMatrix4x4 &matrix, const QgsVector3D &vector )
+{
+  double x, y, z, w;
+
+  x = vector.x() * matrix.m[0][0] +
+      vector.y() * matrix.m[1][0] +
+      vector.z() * matrix.m[2][0] +
+      matrix.m[3][0];
+  y = vector.x() * matrix.m[0][1] +
+      vector.y() * matrix.m[1][1] +
+      vector.z() * matrix.m[2][1] +
+      matrix.m[3][1];
+  z = vector.x() * matrix.m[0][2] +
+      vector.y() * matrix.m[1][2] +
+      vector.z() * matrix.m[2][2] +
+      matrix.m[3][2];
+  w = vector.x() * matrix.m[0][3] +
+      vector.y() * matrix.m[1][3] +
+      vector.z() * matrix.m[2][3] +
+      matrix.m[3][3];
+  if ( w == 1.0f )
+    return QgsVector3D( x, y, z );
+  else
+    return QgsVector3D( x / w, y / w, z / w );
+}
+
+bool QgsMatrix4x4::isIdentity() const
+{
+  if ( m[0][0] != 1.0f || m[0][1] != 0.0f || m[0][2] != 0.0f )
+    return false;
+  if ( m[0][3] != 0.0f || m[1][0] != 0.0f || m[1][1] != 1.0f )
+    return false;
+  if ( m[1][2] != 0.0f || m[1][3] != 0.0f || m[2][0] != 0.0f )
+    return false;
+  if ( m[2][1] != 0.0f || m[2][2] != 1.0f || m[2][3] != 0.0f )
+    return false;
+  if ( m[3][0] != 0.0f || m[3][1] != 0.0f || m[3][2] != 0.0f )
+    return false;
+  return ( m[3][3] == 1.0f );
+}
+
+void QgsMatrix4x4::setToIdentity()
+{
+  m[0][0] = 1.0f;
```

Comment:
```suggestion
  m[0][0] = 1.0;
```

'f' suffix is for 'float', 'l' for 'long double', no suffix for double: https://en.cppreference.com/w/cpp/language/floating_literal